jax/custom_batching.py

```diff
@@ -15,4 +15,5 @@
 from jax._src.custom_batching import (
   custom_vmap as custom_vmap,
   sequential_vmap as sequential_vmap,
+  broadcasting_vmap as broadcasting_vmap,
```

Comment:
why expose this? Every new API that we add has a cost. Please consider not exposing it or do it via jex.

Reply:
I don't think sticking this alone in `jex` is right, but the rest of your point stands.

Stepping back, all of `custom_batching` is part of an effort to expose (effectively) a coherent custom primitive creation interface. It's still up for debate whether that's best done by:
* instantiating something called a custom primitive; or
* attaching custom derivatives, custom batching, custom etc. to a Python callable; or
* anything else.

That overall effort should lead to APIs that either go in `jax.extend` or into `jax`. Note that custom derivatives have long been in `jax`. I think your original question applies to all of `custom_batching` (which is currently in `jax` following custom AD, but not documented as we very slowly "land and iterate").

Separately, whether to add this particular `broadcasting_vmap` convenience to `custom_batching` is its own question. The reason for it is that the general `custom_vmap` API requires a complex implementation (specifically handling reverse-mode AD). Meanwhile we only know of ~2 concrete use cases, equivalent to `sequential_vmap` and `broadcasting_vmap`, for which AD seems easier to handle. So the idea is to expose _only_ the latter two as the API, _instead_ of a general `custom_vmap` function.

Reply:
Do we need to expose that right now or maybe later when we have full fledged custom_batching support?

Reply:
Great points all around!

For the bigger question: I definitely agree that `custom_batching` probably shouldn't be exposed as a top level `jax` API given it's current state of readiness. Maybe it does make sense to migrate it to `jex`, or maybe it's better to just do nothing and wait until we have a better and more complete story around customization.

As for this specific function: I do think that it is useful to expose something like this, and it should probably live in the same place as `sequential_vmap`. I think something like `jax.extend.batching` or `jax.extend.custom_batching` could be a reasonable place, and I don't have strong opinions.

If the consensus is helpers like this and `sequential_vmap` shouldn't be exposed as part of any API, I'm happy to abandon this PR for `broadcasting_vmap` to just exist as an issue comment. No stress either way!

Reply:
I think `convert_to_ufunc` is a clearer name, and is also the driving intention.

Reply:
Re: publish in API. People can grab the decorator from the github issue easily enough when they stumble upon this. _But_, downstream I see a lot more people using `pure_callback` especially in the science community. A great deal of untapped JAX potential was released when you enabled JAX to interoperate with preexisting code. So if you care about this segment you should probably endeavour to support that integration with intention. Huge projects are starting to migrate to JAX where underlying simulation software is in some other language, e.g. FORTRAN, which will now get access to ecosystem of ML and probabilistic programming via JAX. Also, worthy other mention is providing pre-allocated output buffers for these simulations might be highly useful.

Reply:
Thanks for your comments here @Joshuaalbert! The context here is that the code in `jax.custom_batching` is only about 80% operational and completely undocumented, so it's not clear that _this_ is the right place to provide a "public" API like this. (It's also not so obvious that having a function in this undocumented module would have better discoverability than a comment in a GitHub issue.) We have [an official compatibility policy](https://jax.readthedocs.io/en/latest/api_compatibility.html) on all functions that are part of the public API (like this would be), so there's maintenance burden to adding new public functions so we should only do that when we're actually solving the right problem.

You're preaching to the choir on the importance of interoperability so please rest assured that that's a high priority even if this specific function doesn't get merged :D

Reply:
This is great to hear. You know, next week I'll be presenting at MODE conference which this year is basically about using JAX + AD to do experimental design optimisation. If you could drop me a tentative unofficial roadmap for interoperability beween JAX and other codebases that would be really useful. I could include it in my talk. There are many ideas too, like automatic vjp and jvp for linear adjoint pairs of operators. JAXbind is other work focused on this.